internal/pkg/dns/dns.go

```diff
@@ -0,0 +1,254 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package dns provides dns server implementation.
+package dns
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"net"
+	"net/netip"
+	"sync"
+	"time"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/cache"
+	"github.com/coredns/coredns/plugin/pkg/proxy"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
+	"github.com/siderolabs/gen/pair"
+	"github.com/siderolabs/gen/xslices"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/talos/internal/pkg/utils"
+)
+
+// NewRunner creates a new Runner.
+func NewRunner(srv Server, logger *zap.Logger) *Runner {
+	return &Runner{srv: srv, logger: logger, done: make(chan struct{})}
+}
+
+// Runner is a dns server handler.
+type Runner struct {
+	srv    Server
+	logger *zap.Logger
+	done   chan struct{}
+}
+
+// Server is a dns server.
+type Server interface {
+	ListenAndServe() error
+	Shutdown() error
+}
+
+// Run runs dns server.
+func (h *Runner) Run() error {
+	defer close(h.done)
+
+	h.logger.Info("starting dns server")
+
+	err := h.srv.ListenAndServe()
+
+	h.logger.Info("dns server stopped", zap.Error(err))
+
+	return err
+}
+
+// Stop stops dns server. It's safe to call even if server is already stopped. It's not safe to call from multiple
+// goroutines.
+func (h *Runner) Stop() {
+	select {
+	case <-h.done:
+		return
+	default:
+	}
+
+	err := h.srv.Shutdown()
+	if err != nil {
+		h.logger.Warn("error shutting down dns server", zap.Error(err))
+	}
+
+	<-h.done
+}
+
+// Cache is a [dns.Handler] to [plugin.Handler] adapter.
+type Cache struct {
+	cache  *cache.Cache
+	logger *zap.Logger
+}
+
+// NewCache creates a new Cache.
+func NewCache(next plugin.Handler, l *zap.Logger) *Cache {
+	c := cache.NewCache("zones", "view")
+	c.Next = next
+
+	return &Cache{cache: c, logger: l}
+}
+
+// ServeDNS implements [dns.Handler].
+func (c *Cache) ServeDNS(wr dns.ResponseWriter, msg *dns.Msg) {
+	_, err := c.cache.ServeDNS(context.Background(), wr, msg)
+	if err != nil {
+		// we should probably call newProxy.Healthcheck() if there are too many errors
+		c.logger.Warn("error serving dns request", zap.Error(err))
```

Comment:
if the logger here is coming from the controller, it might spam a lot `controller-runtime` logs, probably we'd rather have a separate log for the dns server?

Reply:
Yeah, I thought about this too. I redid initialization in `github.com/siderolabs/talos/internal/app/machined/pkg/runtime/v1alpha2.Controller` so it now have a helper function which is used for both controller logs and logs for dns cache. Im was a bit surprised that `logger.With` method doesn't overwrite existing keys, instead it appends to them.